core/ordersync/ordersync.go

```diff
@@ -0,0 +1,501 @@
+			return
+		}
+		rawReq, err := waitForRequest(s.ctx, stream)
+		if err != nil {
+			log.WithError(err).Warn("waitForRequest returned error")
+			return
+		}
+		log.WithFields(log.Fields{
+			"requester": stream.Conn().RemotePeer().Pretty(),
+		}).Trace("received ordersync request")
+		if rawReq.Type != TypeRequest {
+			log.WithField("gotType", rawReq.Type).Warn("wrong type for Request")
+			s.handlePeerScoreEvent(remotePeerID, psInvalidMessage)
+			return
+		}
+		subprotocol, err := s.GetMatchingSubprotocol(rawReq)
+		if err != nil {
+			log.WithError(err).Warn("GetMatchingSubprotocol returned error")
+			s.handlePeerScoreEvent(remotePeerID, psSubprotocolNegotiationFailed)
+			return
+		}
+		res, err := handleRequestWithSubprotocol(s.ctx, subprotocol, rawReq)
+		if err != nil {
+			log.WithError(err).Warn("subprotocol returned error")
+			return
+		}
+		encodedMetadata, err := json.Marshal(res.Metadata)
+		if err != nil {
+			log.WithError(err).Error("could not encode raw metadata")
+			return
+		}
+		s.handlePeerScoreEvent(remotePeerID, psValidMessage)
+		rawRes := rawResponse{
+			Type:        TypeResponse,
+			Subprotocol: subprotocol.Name(),
+			Orders:      res.Orders,
+			Complete:    res.Complete,
+			Metadata:    encodedMetadata,
+		}
+		if err := json.NewEncoder(stream).Encode(rawRes); err != nil {
+			log.WithFields(log.Fields{
+				"error":     err.Error(),
+				"requester": remotePeerID.Pretty(),
+			}).Warn("could not encode ordersync response")
+			s.handlePeerScoreEvent(remotePeerID, psUnexpectedDisconnect)
+			return
+		}
+		if res.Complete {
+			return
+		}
+	}
+}
+
+// GetOrders iterates through every peer the node is currently connected to
+// and attempts to perform the ordersync protocol. It keeps trying until
+// ordersync has been completed with minPeers, using an exponential backoff
+// strategy between retries.
+func (s *Service) GetOrders(ctx context.Context, minPeers int) error {
+	successfullySyncedPeers := stringset.New()
+
+	for len(successfullySyncedPeers) < minPeers {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		// TODO(albrow): Do this for loop partly in parallel.
+		currentNeighbors := s.node.Neighbors()
+		shufflePeers(currentNeighbors)
+		for _, peerID := range currentNeighbors {
+			if len(successfullySyncedPeers) >= minPeers {
+				return nil
+			}
+			if successfullySyncedPeers.Contains(peerID.Pretty()) {
+				continue
+			}
+
+			log.WithFields(log.Fields{
+				"provider": peerID.Pretty(),
+			}).Trace("requesting orders from neighbor via ordersync")
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+
+			if err := s.getOrdersFromPeer(ctx, peerID); err != nil {
+				log.WithFields(log.Fields{
+					"error":    err.Error(),
+					"provider": peerID.Pretty(),
+				}).Warn("could not get orders from peer via ordersync")
+				continue
+			} else {
+				// TODO(albrow): Handle case where no orders were returned from this
```

Comment:
In line with this, should there be a threshold under which `ordersync` isn't considered a success that is nonzero? My reasoning for including this is that it's possible that a peer has significantly less orders than other peers, and it would make sense to consider that step of `ordersync` a failure. I'm curious to hear whether you think this is necessary considering that we are sampling from multiple peers.

An example heuristic would be "if one of the results of getting orders from a peer contained < 1% of the orders of the other ordersync peers, repeat." This heuristic would necessarily rely on backtracking but would prevent against calling ordersync on a newish peer.

Reply:
I think it is too difficult to come up with a heuristic like this that is accurate, doesn't have any side effects, and is not gameable. IMO it's better to just request orders from more peers to get a better sampling. It's true that some of those peers might also be "new" but statistically it is not very likely that all of them will be. I think I'm going to go ahead and increase the default `minPeers` for now and leave it at that.

We can address this with better subprotocols in the future. The current implementation is more of a stopgap which greatly improves he previous algorithm but has some known flaws. This is one of them.

Reply:
That's reasonable. I think it's possible that we could come up with a working heuristic in future subprotocols if necessary, but I agree that it doesn't seem necessary given a greater amount of sampling.